src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java

```diff
@@ -22,86 +22,178 @@
+        if (priority <= currentPriority) {
+            throw new IllegalStateException("Notifications for an upcoming checkpoint are already in progress (priority "
+                    + currentPriority + "). Please make sure to register this resource earlier or use higher priorty (" + priority + ")");
+        }
     }
 
     @Override
-    public synchronized void beforeCheckpoint(Context<? extends Resource> context) throws CheckpointException {
-        List<R> resources = checkpointQ.entrySet().stream()
-            .sorted(comparator)
-            .map(Map.Entry::getKey)
-            .collect(Collectors.toList());
-
-        CheckpointException exception = new CheckpointException();
-        for (Resource r : resources) {
-            if (FlagsHolder.DEBUG) {
-                System.err.println("jdk.crac beforeCheckpoint " + r.toString());
-            }
+    public void beforeCheckpoint(Context<? extends Resource> context) throws CheckpointException {
+        // If afterRestore is running we need to delay the beforeCheckpoint
+        restoreLock.lock();
+        try {
+            checkpointLock.lock();
             try {
-                r.beforeCheckpoint(this);
-            } catch (CheckpointException e) {
-                for (Throwable t : e.getSuppressed()) {
-                    exception.addSuppressed(t);
+                runBeforeCheckpoint();
+            } finally {
+                if (restoreQ != null) {
+                    Collections.reverse(restoreQ);
                 }
-            } catch (Exception e) {
-                exception.addSuppressed(e);
+                currentPriority = -1;
+                checkpointLock.unlock();
             }
+        } finally {
+            restoreLock.unlock();
+        }
+    }
+
+    private void runBeforeCheckpoint() throws CheckpointException {
+        Map.Entry<R, Long> drained;
+        while ((drained = resourceQueue.poll()) != null) {
+            resources.put(drained.getKey(), drained.getValue());
         }
+        CheckpointException exception = null;
+        TreeMap<Long, List<R>> resources = this.resources.entrySet().stream().collect(
+                TreeMap::new, (m, e) -> m.computeIfAbsent(e.getValue(), p -> new ArrayList<>()).add(e.getKey()), TreeMap::putAll);
+        restoreQ = new ArrayList<>(this.resources.size());
 
-        Collections.reverse(resources);
-        restoreQ = resources;
+        // We cannot simply iterate because we could cause mutations
+        while (!resources.isEmpty()) {
+            var entry = resources.firstEntry();
+            resources.remove(entry.getKey());
+            currentPriority = entry.getKey();
+            for (R r : entry.getValue()) {
+                LoggerContainer.debug("beforeCheckpoint {0}", r);
+                try {
+                    r.beforeCheckpoint(this);
+                    restoreQ.add(r);
+                } catch (CheckpointException e) {
+                    enqueueIfContext(r);
```

Comment:
Ohh, I see the problem

> When Context.beforeCheckpoint throws, invoke Context.afterRestore anyway (otherwise some resources stay in suspended state).

So this is proposed to be handled in the parent context? Have you considered fixing that in the child context, run afterRestore for successfully checkpointed resources, before throwing CheckpointException to the parent context?

Reply:
Right now we run the checkpoint for all resources even if some fail. Are you fine with erroring out on the first exception?

Reply:
Why is that necessary? If we are trying to avoid second afterRestore for sucessfully checkpointed Resources, that would mean that Context.afterRestore is called, that controverts the problem definition.

Assuming afterRestore is not called for the Resource with failed beforeCheckpoint. Suppose we have R1, R2 in a child Context, and R1.beforeCheckpoint succeed, and R2.beforeCheckpoint fails. We can call R1.afterRestore, before throwing from Ctx.beforeCheckpoint. That will restore successfully checkpointed resources.

But I checked the code and it seems that we do call afterRestore regardless of the result of beforeCheckpoint 

https://github.com/openjdk/crac/blob/95394e84683f1a816c0283f8c834072324516fba/src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java#L54

Every Resource available is collected in List resources and then restore queue is exactly the reverse of that list, it does not matter if any of resources throwed CheckpointException. Right?

Reply:
On regular resources `afterRestore` is not called if the resource throws, it does not end up in the restoreQ.

I am a bit worried about calling `R3.bC` that would live in a different context when `R1` is already restored. The components were called in certain order for a reason, and if R3 expect R1 to be checkpointed this could end up badly.
You might argue that R2 is not in checkpointed state anyway and R3 might expect it to be, but at least here one component failed and there is a track of consequences. On the other hand R1 and R3 might be completely unrelevant to R2 and the fact that R2 failed shouldn't interfere.

What is the point of continuing with the checkpoint when we now that it won't happen in the end, with all the suppressed exceptions in the first place? Just to collect all errors at once rather than one-by-one? Is it worth potential false positives?

Reply:
> On regular resources afterRestore is not called if the resource throws, it does not end up in the restoreQ.

I'm puzzled how this is possible in the implementation before the patch. The code says it should
```
        List<R> resources = checkpointQ. ... ;
        for (Resource r : resources) { ... }
        Collections.reverse(resources);
        restoreQ = resources;
```

We'll have to document the expected behavior, but that is another concern.

> I am a bit worried about calling R3.bC that would live in a different context when R1 is already restored.

This is a question of specification. If we set that, it will be adopted by rearranging Contextes and Resources, pretty manageable.

In general, resources should not have a lot of assumptions about other resources. Otherwise, they should directly refer to each other, rathen than relying on the Context. That will be more straightforward and less error-prone.

> What is the point of continuing with the checkpoint when we now that it won't happen in the end, with all the suppressed exceptions in the first place? Just to collect all errors at once rather than one-by-one? Is it worth potential false positives?

Correct, to run all registered code, that may also have some side-effects (unpreferable, but possible). Having that checkpoint won't happen anyway, false positive does not look a problem.

Reply:
True, checking again before the patch it was really calling it on resources even after failure - I had "fixed" that probably way before and remembered it that way. When a code is throwing it is aware of that, so it should either revert automatically or be considered completely broken. I would bring an analogy with a locking code - `lock()` method is executed outside of the try-finally block so if it fails it should keep the lock unlocked.

The R1 and R3 issue - it is fixable, true, but it's the infra code that introduced the failure. User sees both R2 and R3 failures, decides to investigate R3 only to find that it was a red herring.

> resources should not have a lot of assumptions about other resources

JDK code is a direct contradiction to that. We shall not expect user code to be any different.

> false positive does not look a problem

Red herrings are UX problem. And it is a problem when the failure leaves the JVM in a state that does not allow inspection - not sure why but I had cases where neither `jstack` nor `kill -3` worked.